mkdocs_simple_plugin/plugin.py

```diff
@@ -3,13 +3,139 @@
 from mkdocs.config import config_options
 from mkdocs import utils
 
+from enum import Enum, auto
 import fnmatch
 import os
+import re
 import shutil
 import sys
 import tempfile
 
 
+class LazyFile:
+    """
+    Like a file object, but only ever creates the directory and opens the
+    file if a non-empty string is written.
+    """
+    def __init__(self, direc, name):
+        self.fdir = direc
+        self.fname = name
+        self.fobj = None
+
+    def write(self, arg):
+        if arg == '':
+            return
+        if self.fobj is None:
+            os.makedirs(self.fdir, exist_ok=True)
+            self.fobj = open(f"{self.fdir}/{self.fname}", 'w')
+        self.fobj.write(arg)
+
+    def close(self):
+        if self.fobj is not None:
+            self.fobj.close()
+
+
+class ExState(Enum):
+    Waiting = auto()
+    Writing = auto()
+    Paused = auto()
+
+
+class StreamExtract:
+    """
+    Instantiating a StreamExtract object copies _instream_ to _outstream_,
+    extracting portions of the instream as specified by the keyword arguments
+    to the constructor as documented for the "semiliterate" parameter of the
+    `simple` plugin.
+    """
+    def __init__(self, instream, outstream,
+                 start=None, end=None, pause=None, resume=None, replace=[],
+                 incroot=None, **ignore_other_kwargs):
+        self.ins = instream
+        self.outs = outstream
+        self.start = (start is not None) and re.compile(start)
+        self.end = (end is not None) and re.compile(end)
+        self.pause = (pause is not None) and re.compile(pause)
+        self.resume = (resume is not None) and re.compile(resume)
+        self.replace = []
+        for item in replace:
+            if isinstance(item, str):
+                self.replace.append(re.compile(item))
+            else:
+                self.replace.append((re.compile(item[0]), item[1]))
+        self.incroot = incroot
+        if self.start and self.pause and not self.resume:
+            self.resume = self.start
+        self.wroteSomething = False
+        self.state = ExState.Waiting if self.start else ExState.Writing
+        self.extract()
+
+    def writeout(self, txt):
+        self.outs.write(txt)
+        if txt:
+            self.wroteSomething = True
+
+    def checkpat(self, pat, line, emitlast=True):
```

Comment:
Is 'pat' supposed to be short for 'pattern'?  Prefer whole words in function names.

Reply:
Gotcha, will rename when I revise the PR.